extra/tensors/tensors.factor

```diff
@@ -0,0 +1,257 @@
+METHOD: t* { number tensor } swap [ * ] curry t-uop ;
+
+! Divide two tensors or a tensor and a scalar
+multi-methods:GENERIC: t/ ( x y -- tensor )
+METHOD: t/ { tensor tensor } [ / ] t-bop ;
+METHOD: t/ { tensor number } [ / ] curry t-uop ;
+METHOD: t/ { number tensor } swap [ swap / ] curry t-uop ;
+
+! Divide two tensors or a tensor and a scalar
+multi-methods:GENERIC: t% ( x y -- tensor )
+METHOD: t% { tensor tensor } [ mod ] t-bop ;
+METHOD: t% { tensor number } [ mod ] curry t-uop ;
+METHOD: t% { number tensor } swap [ swap mod ] curry t-uop ;
+
+<PRIVATE
+
+! Check that the tensor has an acceptable shape for matrix multiplication
+: check-matmul-shape ( tensor1 tensor2 -- )
+    [let [ shape>> ] bi@ :> shape2 :> shape1
+    ! Check that the matrices can be multiplied
+    shape1 last shape2 [ length 2 - ] keep nth =
+    ! Check that the other dimensions are equal
+    shape1 2 head* shape2 2 head* = and
+    ! If either is false, raise an error
+    [ shape1 shape2 shape-mismatch-error ] unless ] ;
+
+! Slice out a row from the array
+: row ( arr n i p -- slice )
+    ! Compute the starting index
+    / truncate dupd *
+    ! Compute the ending index
+    swap over +
+    ! Take a slice
+    rot <slice> ;
+
+! Perform matrix multiplication muliplying an
+! mxn matrix with a nxp matrix
+TYPED:: 2d-matmul ( vec1: slice vec2: slice res: slice n: number p: number -- )
+    ! For each element in the range, we want to compute the dot product of the
+    ! corresponding row and column
+    res
+    [   >fixnum
+        ! Get the row
+        [ [ vec1 n ] dip p row ]
+        ! Get the column
+        ! [ p mod vec2 swap p every ] bi
+        [ p mod f p vec2 <step-slice> ] bi
+        ! Take the dot product
+        [ * ] [ + ] 2map-reduce
+    ]
+    map! drop ;
+
+PRIVATE>
+
+
+! Perform matrix multiplication muliplying an
+! ...xmxn matrix with a ...xnxp matrix
+TYPED:: matmul ( tensor1: tensor tensor2: tensor -- tensor3: tensor )
+    ! First check the shape
+    tensor1 tensor2 check-matmul-shape
+
+    ! Now save all of the sizes
+    tensor1 shape>> unclip-last-slice :> n
+    unclip-last-slice :> m :> top-shape
+    tensor2 shape>> last :> p
+    top-shape product :> rest
+
+    ! Now create the new tensor with { 0 ... m*p-1 } repeating
+    top-shape { m p } append naturals m p * t% :> tensor3
+
+    ! Now update the tensor3 to contain the multiplied matricies
+    rest [0,b)
+    [
+        :> i
+        ! First make vec1
+        m n * i * dup m n * + tensor1 vec>> <slice>
+        ! Now make vec2
+        n p * i * dup n p * + tensor2 vec>> <slice>
+        ! Now make the resulting vector
+        m p * i * dup m p * + tensor3 vec>> <slice>
+        ! Push n and p and multiply the clices
+        n p 2d-matmul
+        0
+    ] map drop
+    tensor3 ;
+
+<PRIVATE
+! helper for transpose: gets the turns a shape into a list of things
+! by which to multiply indices to get a full index
+: get-mults ( shape -- seq )
```

Comment:
This could use a better name maybe.

Other ideas for the implementation:

```
: get-mults ( shape -- seq )
    rest-slice <reversed> cum-product { 1 } prepend ;
```

I suspect this and ``trans-index`` are pretty slow since you're making these multiplier arrays for each index operation.

Reply:
And the multiplier arrays are constant with shape, not index.